server/mqtt_ha.go

```diff
@@ -0,0 +1,114 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/evcc-io/evcc/core/site"
+)
+
+type HADeviceDef struct {
+	Identifiers  string `json:"identifiers"`
+	Manufacturer string `json:"manufacturer"`
+	Name         string `json:"name"`
+}
+
+type HAEntityDef struct {
+	Device   HADeviceDef `json:"device"`
+	Name     string      `json:"name"`
+	UniqueId string      `json:"unique_id"`
+
+	StateTopic        string `json:"state_topic,omitempty"`
+	CommandTopic      string `json:"command_topic,omitempty"`
+	AvailabilityTopic string `json:"availability_topic,omitempty"`
+
+	Options []string `json:"options,omitempty"`
+
+	Min float64 `json:"min"`
+	Max float64 `json:"max"`
+
+	StateClass string `json:"state_class,omitempty"`
+}
+
+func (m *MQTT) haPublishBaseDeviceDef(site site.API, loadPoint *int, valueName string) HAEntityDef {
+	siteId := strings.ReplaceAll(strings.ToLower(site.Name()), " ", "_") // TODO find better site ID
+	uid := "evcc_" + siteId + "_"
+	if loadPoint != nil {
+		uid += fmt.Sprintf("lp%d_", *loadPoint+1)
+	}
+	uid += valueName
+
+	var name string
+	if loadPoint == nil {
+		name = fmt.Sprintf("EVCC %s %s", site.Name(), valueName)
+	} else {
+		name = fmt.Sprintf("EVCC %s Loadpoint %d %s", site.Name(), *loadPoint+1, valueName)
+	}
+
+	return HAEntityDef{
+		Device: HADeviceDef{
+			Identifiers:  fmt.Sprintf("evcc_%s", siteId),
+			Manufacturer: "EVCC",
+			Name:         site.Name(),
+		},
+		Name:              name,
+		UniqueId:          uid,
+		AvailabilityTopic: m.root + "/status",
+	}
+}
+
+func (m *MQTT) haPublishDiscoverSensors(site site.API, loadPoint *int, valueName, stateTopic string, val interface{}) {
+	if _, ok := m.haKnownSensors[stateTopic]; ok {
+		return
+	}
+
+	entityDef := m.haPublishBaseDeviceDef(site, loadPoint, valueName)
+	entityDef.StateTopic = stateTopic
+
+	switch val.(type) {
+	case float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, []float64:
+		entityDef.StateClass = "measurement"
+	}
+	topic := "homeassistant/sensor/" + entityDef.UniqueId + "/config"
+
+	jsonData, _ := json.MarshalIndent(entityDef, "", "  ")
+	token := m.Handler.Client.Publish(topic, m.Handler.Qos, true, jsonData)
```

Comment:
Muss das denn alles teil des mqtt Objektes sein oder würde es nicht reichen, `handler` als Parameter zu bekommen? Dann könnte das alles in einen eigenen `homeassisstant` Ordner.

Reply:
das `AvailabilityTopic` hohle ich mir basierend auf dem `root` vom MQTT Objekt.
Schön ist das auch nicht da das jetzt an 2 Stellen definiert wird